lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+      .Default([&](Type t) { return t; });
+}
+
+static Type toESIHWType(Type t) {
+  auto *ctx = t.getContext();
+  Type outType =
+      llvm::TypeSwitch<Type, Type>(t)
+          .Case<ValueType>([&](auto vt) {
+            return esi::ChannelType::get(ctx, toHWType(vt.getInnerTypes()));
+          })
+          .Case<TokenType>([&](auto tt) {
+            return esi::ChannelType::get(ctx,
+                                         IntegerType::get(tt.getContext(), 0));
+          })
+          .Default([](auto t) { return toHWType(t); });
+
+  return outType;
+}
+
+namespace {
+
+/// Shared state used by various functions; captured in a struct to reduce the
+/// number of arguments that we have to pass around.
+struct DCLoweringState {
+  ModuleOp parentModule;
+  NameUniquer nameUniquer;
+};
+
+/// A type converter is needed to perform the in-flight materialization of "raw"
+/// (non-ESI channel) types to their ESI channel correspondents. This comes into
+/// effect when backedges exist in the input IR.
+class ESITypeConverter : public TypeConverter {
+public:
+  ESITypeConverter() {
+    addConversion([](Type type) -> Type { return toESIHWType(type); });
+    addConversion([](esi::ChannelType t) -> Type { return t; });
+
+    addTargetMaterialization(
+        [](mlir::OpBuilder &builder, mlir::Type resultType,
+           mlir::ValueRange inputs,
+           mlir::Location loc) -> std::optional<mlir::Value> {
+          if (inputs.size() != 1)
+            return std::nullopt;
+          return inputs[0];
+        });
+
+    addSourceMaterialization(
+        [](mlir::OpBuilder &builder, mlir::Type resultType,
+           mlir::ValueRange inputs,
+           mlir::Location loc) -> std::optional<mlir::Value> {
+          if (inputs.size() != 1)
+            return std::nullopt;
+          return inputs[0];
+        });
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// HW Sub-module Related Functions
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Input handshakes contain a resolved valid and (optional )data signal, and
+/// a to-be-assigned ready signal.
+struct InputHandshake {
+  Value channel;
+  Value valid;
+  std::shared_ptr<Backedge> ready;
+  Value data;
+};
+
+/// Output handshakes contain a resolved ready, and to-be-assigned valid and
+/// (optional) data signals.
+struct OutputHandshake {
+  Value channel;
+  std::shared_ptr<Backedge> valid;
+  Value ready;
+  std::shared_ptr<Backedge> data;
+};
+
+///  Directly connect an input handshake to an output handshake
+static void connect(InputHandshake &input, OutputHandshake &output) {
+  output.valid->setValue(input.valid);
+  input.ready->setValue(output.ready);
+}
+
+/// A helper struct that acts like a wire. Can be used to interact with the
+/// RTLBuilder when multiple built components should be connected.
+struct HandshakeWire {
```

Comment:
I cannot find any usage of this class, is this a remnant?

Reply:
it is!